packages/playwright-test/src/runner.ts

```diff
@@ -270,74 +314,84 @@ export class Runner {
   }
 
   private async _run(options: RunOptions): Promise<FullResult> {
-    const testFileFilters = options.testFileFilters || [];
-    const filesByProject = await this._collectFiles(testFileFilters, options.projectFilter);
-
-    const allTestFiles = new Set<string>();
-    for (const files of filesByProject.values())
-      files.forEach(file => allTestFiles.add(file));
-
     const config = this._loader.fullConfig();
-
     const fatalErrors: TestError[] = [];
+    // Each entry is an array of test groups that can be run concurrently. All
+    // test groups from the previos entries must finish before entry starts.
+    const concurrentTestGroups = [];
+    const rootSuite = new Suite('', 'root');
+    const runPhases = this._collectRunPhases(options);
+    assert(runPhases.length > 0);
+    for (const { projectFilter, testFileFilters } of runPhases) {
```

Comment:
`3. Filter tests to respect line/column filter.` - unclear how this should work with groups.

Reply:
Added TODO.